### DISCUSSION TITLE
Fully Automated Releases

### DISCUSSION THREAD

**Discussion by shadycuz** [OpenBB-finance/OpenBB#91] [Ideas]
This project has already seen a ton of interest in spite of the difficulty in getting it installed. I expect in the future that the amount of contributors to skyrocket judging from the forks and the stars. To maintain velocity and operational excellence we are going to need end to end automation. I hope this discussion helps facilitate both the design and the implementation. 

Goals:
- Publish Release Candidates to Pypi
- Publish Releases to Pypi
- Create Tags for all releases
- Pubilsh Docker Images for all releases

The PR workflow I propose is:
1. Create a branch from "develop"
2. Use `pre-commit` to run CI/CD locally for every commit
3. Create PR back to develop (more CI/CD stuff runs)
4. Label the PR for the type of change (major, minor, patch)

When a merge happens into `develop` we will want to:

1. Create a release candidate tag
2. Create a Github Draft Release
3. Publish to Pypi with the release candidate tag.
4. Publish to DockerHub

The plan is to allow multiple merges into `develop` and continue to create release candidates until we are ready to actually release a new version by opening a PR from `develop` into `main`. 

When a PR is merged to `main`:
1. Create a release tag
2. Convert Draft Release to an actual Release.
3. Publish to Pypi
4. Publish to DockerHub

If we can build a pipeline to do all the steps we would be in a really great position. The `develop` section is actually the hardest. Lots of github actions can help us create tags, create gthub releases etc but not all of them do tags and stuff in the format that pypi specifies. Thankfully we can use `poetry` to help with most of that. It has some limitations that I have outlined [here](https://github.com/python-poetry/poetry/issues/3752).

Say the version of the application is `1.0.0`.  If a `patch` PR is merged into `develop` then we need to create a `1.0.1-alpha` tag. If a second patch is merged then we need to update the tag to `1.0.1-alpha.1`. If a `minor` PR is merged into develop` then we need to create `1.1.0-alpha` tag. If another `patch` PR is merged the tag should be `1.1.0-alpha.1`. 

It seems pretty straight forward but it will be a complicated workflow to implement. I have implemented about 85% of this for other projects in the past and I think 100% e2e is doable nowadays.

Feel free to share your thoughts but we need both the docker and pip stuff now. So I will begin implementing parts of this in the next few days. 

@DidierRLopes @cclauss @akx

**Comment by DidierRLopes** [comment]
That sounds good to me!

**Reply by shadycuz** [reply]
@aia

**Comment by aia** [comment]
Sounds great!
